File: examples/go/anti_patterns.go
// Anti-patterns vs correct patterns for FC/IS architecture.
package examples

import (
	"encoding/json"
	"os"
)

// ============================================================
// 1. Raw data in core
// ============================================================

// ANTI-PATTERN: Core function receives raw JSON map
func CalculateShippingBad(raw map[string]interface{}) float64 {
	weight := raw["weight"].(float64) // panics if missing or wrong type
	if weight < 10 {
		return weight * 0.5
	}
	return weight * 0.3
}

// CORRECT: Core function receives parsed domain type
type ShippingOrder struct {
	WeightKg float64
}

func CalculateShipping(order ShippingOrder) float64 {
	if order.WeightKg < 10 {
		return order.WeightKg * 0.5
	}
	return order.WeightKg * 0.3
}

// ============================================================
// 2. Validation instead of parsing
// ============================================================

// ANTI-PATTERN: Returns bool, caller still uses raw value
func IsValidQuantity(q int) bool {
	return q > 0
}

// CORRECT: Returns parsed type or error
type Quantity struct{ Value int }

func ParseQuantity(raw int) (Quantity, error) {
	if raw < 1 {
		return Quantity{}, &ParseError{Field: "quantity", Message: "must be positive"}
	}
	return Quantity{Value: raw}, nil
}

// ============================================================
// 3. I/O in core
// ============================================================

// ANTI-PATTERN: Core reads environment directly
func GetDiscountBad(price float64) float64 {
	rate := os.Getenv("DISCOUNT_RATE") // I/O in core!
	_ = rate
	return price * 0.9
}

// CORRECT: Core receives rate as parameter
func GetDiscount(price float64, discountRate float64) float64 {
	return price * (1 - discountRate)
}

// ============================================================
// 4. Untested boundaries
// ============================================================

// ANTI-PATTERN: No parsing, just unmarshal and hope
func HandleOrderBad(data []byte) float64 {
	var order map[string]interface{}
	json.Unmarshal(data, &order) // error ignored
	return CalculateShippingBad(order)
}

// CORRECT: Parse at boundary, handle errors, pass typed value to core
func HandleOrderCorrect(data []byte) (float64, error) {
	var raw struct {
		WeightKg *float64 `json:"weight_kg"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return 0, err
	}
	if raw.WeightKg == nil || *raw.WeightKg < 0 {
		return 0, &ParseError{Field: "weight_kg", Message: "required non-negative number"}
	}
	return CalculateShipping(ShippingOrder{WeightKg: *raw.WeightKg}), nil
}

File: examples/rust/anti_patterns.rs
//! Anti-patterns vs correct patterns for FC/IS architecture.
#![allow(dead_code)]

use std::collections::HashMap;

// ============================================================
// 1. Raw data in core
// ============================================================

// ANTI-PATTERN: Core function receives raw HashMap
fn calculate_shipping_bad(raw: &HashMap<String, String>) -> f64 {
    let weight: f64 = raw.get("weight").unwrap().parse().unwrap(); // panics!
    if weight < 10.0 { weight * 0.5 } else { weight * 0.3 }
}

// CORRECT: Core function receives parsed domain type
struct ShippingOrder {
    weight_kg: f64,
}

fn calculate_shipping(order: &ShippingOrder) -> f64 {
    if order.weight_kg < 10.0 { order.weight_kg * 0.5 } else { order.weight_kg * 0.3 }
}

// ============================================================
// 2. Validation instead of parsing
// ============================================================

// ANTI-PATTERN: Returns bool, caller still uses raw value
fn is_valid_quantity(q: i32) -> bool {
    q > 0
}

// CORRECT: Returns parsed type or error
struct Quantity(u32);

fn parse_quantity(raw: i32) -> Result<Quantity, String> {
    if raw < 1 {
        return Err(format!("quantity must be positive, got {raw}"));
    }
    Ok(Quantity(raw as u32))
}

// ============================================================
// 3. Panics as flow control
// ============================================================

// ANTI-PATTERN: Using unwrap/panic for expected failure cases
fn parse_config_bad(raw: &str) -> u32 {
    raw.parse().unwrap() // panics on invalid input
}

// CORRECT: Return Result for expected failures
fn parse_config(raw: &str) -> Result<u32, String> {
    raw.parse().map_err(|_| format!("invalid config value: {raw:?}"))
}

// ============================================================
// 4. Mutable global state in core
// ============================================================

// ANTI-PATTERN: Core mutates global state
static mut COUNTER: u64 = 0;

fn next_id_bad() -> u64 {
    unsafe {
        COUNTER += 1;
        COUNTER
    }
}

// CORRECT: Core receives ID as parameter
fn create_with_id(name: &str, id: u64) -> (u64, String) {
    (id, name.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_correct_patterns() {
        assert!((calculate_shipping(&ShippingOrder { weight_kg: 5.0 }) - 2.5).abs() < 0.01);
        assert!(parse_quantity(5).is_ok());
        assert!(parse_quantity(-1).is_err());
        assert!(parse_config("42").is_ok());
        assert!(parse_config("abc").is_err());
    }
}
